src/main/groovy/com/pstehlik/groovy/gelf4j/appender/Gelf4JAppender.groovy

```diff
@@ -76,83 +72,33 @@ extends AppenderSkeleton {
-    }
 
-    if (layout == null || layout.ignoresThrowable() || !fullMessage.length()) {
-      throwableAsString = throwableAsString?:loggingEvent.getThrowableStrRep()
-      if (throwableAsString != null) {
-        int len = throwableAsString.length
-        if(fullMessage.length()){ // newline after the 'original log message' to have nicer formatting
-          fullMessage.append Layout.LINE_SEP
-        }
-        for (int i = 0; i < len && i <  this.getMaxLoggedLines(); i++) {
-          fullMessage.append throwableAsString[i]
-          fullMessage.append Layout.LINE_SEP
-        }
-      }
-    }
+    GelfMessageBuilder messageBuilder = new GelfMessageBuilder(this, loggingEvent)
+    messageBuilder.setField('host', loggingHostName)
+    messageBuilder.setAdditionalField('facility', facility ?: 'GELF')
 
-    if(!fullMessage.length()){ //failsave to prevent a 'null' or empty message even though it should(!) not happen
-      fullMessage.append loggingEvent.message as String
-    }
-    String shortMessage = fullMessage.toString()
-    if (shortMessage.length() > SHORT_MESSAGE_LENGTH) {
-      shortMessage = shortMessage.substring(0, SHORT_MESSAGE_LENGTH)
-    }
-    int gelfTimeStamp = Math.floor(loggingEvent.getTimeStamp() / 1000)
-    def gelfMessage = [
-      "facility": facility ?: 'GELF',
-      "file": '',
-      "full_message": fullMessage.toString(),
-      "host": loggingHostName,
-      "level": "${loggingEvent.getLevel().getSyslogEquivalent()}" as String,
-      "line": '',
-      "short_message": shortMessage,
-      "timestamp": gelfTimeStamp as String,
-      "version": GELF_VERSION
-    ]
-    //only set location information if configured
-    if (includeLocationInformation) {
-      gelfMessage.file = loggingEvent.getLocationInformation().fileName
-      gelfMessage.line = loggingEvent.getLocationInformation().lineNumber
-    }
     //add additional fields and prepend with _ if not present already
-    if (additionalFields != null) {
-      additionalFields.each {
-        String key = it.key
-        if (!key.startsWith('_')) {
-          key = '_' + key
-        }
-        //skip additional field called '_id'
-        if (key != '_id') {
-          gelfMessage[key] = it.value as String
-        }
-      }
+    additionalFields.each {
+      messageBuilder.setAdditionalField(it.key as String, it.value as String)
+    }
+
+    String ndc = loggingEvent.getNDC()
+    if (ndc) {
+      messageBuilder.setAdditionalField('ndc', ndc)
     }
 
-    if (mdcFields != null) {
-      mdcFields.each {
-        def mdcValue = loggingEvent.getMDC(it)
-        if (mdcValue != null) {
-          gelfMessage['_' + it] = mdcValue as String
-        }
+    mdcFields.each {
+      def mdcValue = loggingEvent.getMDC(it as String)
+      if (mdcValue != null) {
+        messageBuilder.setAdditionalField(
+          'mdc_' + ((it as String) =~ /([A-Z])/).replaceAll('_$1').toLowerCase(),
```

Comment:
I don't love the idea of running a regex on every field prior to sending them off to graylog, could we do this translation through a map at initialization time?